### ISSUE TITLE
Intermittent failures - TestGemLocalRemoteOptions#test_clear_sources_option_idiot_proof

### ISSUE THREAD

**Issue by MSP-Greg** [rubygems/rubygems#2114]
The above method (shown [here](https://github.com/rubygems/rubygems/blob/ff6ca04a6cbe8c2c137da398b64a4b9b1ecef419/test/rubygems/test_gem_local_remote_options.rb#L43-L49)), intermittently fails.

The last time it showed the following:
```
  1) Failure:
TestGemLocalRemoteOptions#test_clear_sources_option_idiot_proof [C:/projects/rubygems/test/rubygems/test_gem_local_remote_options.rb:48]:
--- expected
+++ actual
@@ -1 +1,3 @@
-["https://rubygems.org/"]
+#<Gem::SourceList:0xXXXXXX
+ @sources=
+  [[Remote: http://test.example.com/ API URI: http://test.example.com/]]>
```

This was during a test run using my fork, but only once, and I did a few runs with several ruby version) tonite. It probably passed 20 or more times, and failed once.

But, my ruby-loco trunk build is run three times a day, and all tests run parallel.  This test often fails during parallel tests, but (as I recall) has always passed on retry.  Obviously, ruby/ruby runs lots of tests on Travis & Appveyor, but they don't log it to the level needed to see this issue.

Anyone have any idea?  I haven't looked thru the code to see setup, teardown, the three methods in the test, etc...  Thanks, Greg

BTW, this is what I originally came to 'issues' for...

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by MSP-Greg** [comment]
I believe I've tracked down what the two possible causes are, but I don't know what the correct solution is.

1. `Gem.default_sources` and `Gem.sources` are doc'd as being arrays.  There seems to be code that allows their members to be either URI type objects or strings.  I'm not sure if that is correct, If it is, some tests methods require one or the other to pass, or, a means needs to be provided to compare arrays composed of either.

2. The [`Gem.sources`](https://github.com/rubygems/rubygems/blob/814ea7a4a0098966dd7ecd2d9cad3fbf1615f2e0/lib/rubygems.rb#L999-L1002) code returns a [`Gem::SourceList`](https://github.com/rubygems/rubygems/blob/814ea7a4a0098966dd7ecd2d9cad3fbf1615f2e0/lib/rubygems/source_list.rb), which, although it has methods it shares with Array, it is not, especially when on the RHS of an equality.

I believe these combine to cause the intermittent failures I've seen.  Again, don't know what to change...

So much for taking a short lunch break.., Thanks, & Happy Holidays / Merry Christmas, etc, Greg

**Comment by MSP-Greg** [comment]
Found the issue.  Will post PR when intermittent failures prove to have been fixed.

**Comment by MSP-Greg** [comment]
Sorry, I decided that I don't know RubyGems well enough, so I've got a question.

The code for test in question is
```ruby
  def test_clear_sources_option_idiot_proof
    spec_fetcher

    @cmd.add_local_remote_options
    @cmd.handle_options %W[--clear-sources]
    assert_equal Gem.default_sources, Gem.sources
  end
```

The issue centers on `configuration.sources`.  The above code calls `handle_options`, which calls the block in `add_clear_sources_option` (first code snippet below.)  This block clears `Gem.sources`, but does not clear `configuration.sources`, which is used in the bottom code snippet to reload sources.

So, should `configuration.sources` be cleared with `--clear-sources`, or should it be cleared in the test?

```ruby
# Code for Gem::LocalRemoteOptions#add_clear_sources_option
def add_clear_sources_option
  add_option(:"Local/Remote", '--clear-sources',
             'Clear the gem sources') do |value, options|

    Gem.sources = nil
    options[:sources_cleared] = true
  end
end
```



```ruby
# Code for Gem.sources
def self.sources
  source_list = configuration.sources || default_sources
  @sources ||= Gem::SourceList.from(source_list)
end
```

As mentioned above, I've ran trunk builds hundreds of time, and all the recent builds show a build message with tests that failed during parallel testing (which almost always pass on retry).  I've also been testing on Appveyor with my fork lately, and this test is one two intermittent test problems.

**Comment by segiddins** [comment]
My guess is, that in the tests `configuration.sources` will be empty anyways? But I'm not sure why the test would only ever fail intermittently.

**Comment by MSP-Greg** [comment]
It should be, but since the order of the tests is random, there must be another test affecting it.  I checked today, and in 54 trunk builds this month (that I've saved), this test failed 28 times during parallel testing.

So I assume it's a test issue, or, that it should not be cleared in the block of `Gem::LocalRemoteOptions#add_clear_sources_option`?

Hence, should it be set to nil after `@cmd.handle_options %W[--clear-sources]` in the test?

BTW, I didn't notice the `to_ary` method in `SouceList`.  Regardless, the object mismatch looks odd when minitest reports the error.

Thanks, Greg

**Comment by MSP-Greg** [comment]
@segiddins

Throughout the day, I ran several builds on my Appveyor fork.  Setting `configuration.sources = nil` seems to have solved the issue.  But, although I had five builds pass with all versions (including trunk & 193), the sixth build had one ruby version fail on `test/rubygems/test_gem_commands_owner_command.rb`.  I've seen that before (mentioned above).

Closing this for good.  Once I get that sorted out, I'll update the open PR.  Thanks, Greg
